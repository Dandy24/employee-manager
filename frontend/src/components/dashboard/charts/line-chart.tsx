import { observer } from 'mobx-react-lite';
import Title from 'antd/lib/typography/Title';
import { CartesianGrid, Legend, Line, LineChart, ResponsiveContainer, Tooltip, XAxis, YAxis } from 'recharts';
import { Col } from 'antd';
import React from 'react';
import { HoursTypeGraphDataInterface } from '../../../models/interfaces/graph-data-interface';
import moment from 'moment';

export interface LineChartProps {
    title: string;
    data: HoursTypeGraphDataInterface[];
    dataKey1: string;
    dataKey2: string;
    dataName1: string;
    dataName2: string;
    xAxisKey: string;
}

export const MyLineChart: React.FC<LineChartProps> = observer((props: LineChartProps) => {
    const { title, data, dataKey1, dataName1, dataKey2, dataName2, xAxisKey } = props;

    return (
        <Col span={12} style={{ height: '90%' }}>
            <Title level={2} style={{ textAlign: 'center' }}>
                {title}
            </Title>
            <ResponsiveContainer id={'line-chart'} width="100%" height="90%">
                <LineChart width={600} height={250} data={data} margin={{ top: 5, right: 30, left: 20, bottom: 5 }}>
                    <CartesianGrid strokeDasharray="4 4" />
                    <XAxis
                        dataKey={xAxisKey}
                        tickMargin={10}
                        tickFormatter={(value) => moment(value, 'YYYY-MM-DD').format('MMMM YY')}
                    />
                    <YAxis />
                    <Tooltip
                        labelFormatter={(label: string) => <p>{moment(label, 'YYYY-MM-DD').format('MMMM YY')}</p>}
                        formatter={(amount) => `${amount} hod`}
                    />
                    <Legend id={'line-chart-legend'} />
                    <Line
                        type="monotone"
                        dataKey={dataKey1}
                        name={dataName1}
                        stroke="#8884d8"
                        id={`line-chart-${dataKey1}-line`}
                    />
                    <Line
                        type="monotone"
                        dataKey={dataKey2}
                        name={dataName2}
                        stroke="#82ca9d"
                        id={`line-chart-${dataKey2}-line`}
                    />
                </LineChart>
            </ResponsiveContainer>
        </Col>
    );
});
