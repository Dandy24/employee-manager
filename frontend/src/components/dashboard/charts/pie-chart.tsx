import { observer } from 'mobx-react-lite';
import Title from 'antd/lib/typography/Title';
import { Cell, Legend, Pie, PieChart, ResponsiveContainer, Tooltip } from 'recharts';
import { Col } from 'antd';
import React from 'react';
import { GraphDataInterface } from '../../../models/interfaces/graph-data-interface';

export interface PieChartProps {
    title: string;
    data: GraphDataInterface[];
    dataKey: string;
    dataName: string;
    colors: string[];
}

export const MyPieChart: React.FC<PieChartProps> = observer((props: PieChartProps) => {
    const { title, data, dataKey, dataName, colors } = props;

    return (
        <Col span={10} offset={1} style={{ height: '90%' }}>
            <Title level={2} style={{ textAlign: 'center' }}>
                {title}
            </Title>
            <ResponsiveContainer width="100%" height="85%" id={'pie-chart'}>
                <PieChart width={300} height={200} margin={{ top: 0, right: 0, bottom: 50, left: 0 }}>
                    <Pie
                        data={data}
                        cy={130}
                        innerRadius={70}
                        outerRadius={100}
                        fill="#8884d8"
                        paddingAngle={3}
                        dataKey={dataKey}
                        nameKey={dataName}
                        isAnimationActive={false}
                        label={({ percent }) => `${(percent * 100).toFixed(2)}%`}
                    >
                        {data.map((entry, index) => (
                            <Cell
                                id={`pie-chart-${entry.name.toLowerCase().replace(' ', '-')}-sector`}
                                key={`cell-${index}`}
                                fill={colors[index % colors.length]}
                            />
                        ))}
                    </Pie>
                    <Legend wrapperStyle={{ bottom: '8%' }} id={'pie-chart-legend'} />
                    <Tooltip formatter={(amount) => `${amount} hours`} />
                </PieChart>
            </ResponsiveContainer>
        </Col>
    );
});
