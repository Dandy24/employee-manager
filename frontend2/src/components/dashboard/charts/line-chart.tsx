import { observer } from 'mobx-react-lite';
import Title from 'antd/lib/typography/Title';
import { CartesianGrid, Legend, Line, LineChart, ResponsiveContainer, Tooltip, XAxis, YAxis } from 'recharts';
import { Col } from 'antd';
import React from 'react';
import { HoursTypeGraphDataInterface } from '../../../models/interfaces/graph-data-interface';
import moment from 'moment';

export interface LineChartProps {
    title: string;
    data: HoursTypeGraphDataInterface[];
    dataKey1: string;
    dataKey2: string;
    dataName1: string;
    dataName2: string;
    xAxisKey: string;
}

export const MyLineChart: React.FC<LineChartProps> = observer((props: LineChartProps) => {
    const { title, data, dataKey1, dataName1, dataKey2, dataName2, xAxisKey } = props;

    return (
        <Col span={15} style={{ height: '90%' }}>
            <Title level={2} style={{ textAlign: 'center' }}>
                {title}
            </Title>
            <ResponsiveContainer width="100%" height="90%">
                <LineChart width={600} height={250} data={data} margin={{ top: 5, right: 30, left: 20, bottom: 5 }}>
                    <CartesianGrid strokeDasharray="3 3" />
                    <XAxis
                        dataKey={xAxisKey}
                        tickMargin={10}
                        tickFormatter={(value) => moment(value, 'YYYY-MM-DD').format('MMMM YY')}
                    />
                    <YAxis />
                    <Tooltip />
                    <Legend />
                    <Line type="monotone" dataKey={dataKey1} name={dataName1} stroke="#8884d8" />
                    <Line type="monotone" dataKey={dataKey2} name={dataName2} stroke="#82ca9d" />
                </LineChart>
            </ResponsiveContainer>
        </Col>
    );
});
