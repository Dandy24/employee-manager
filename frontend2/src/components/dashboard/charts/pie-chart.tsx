import { observer } from 'mobx-react-lite';
import Title from 'antd/lib/typography/Title';
import { Cell, Legend, Pie, PieChart, ResponsiveContainer, Tooltip } from 'recharts';
import { Col } from 'antd';
import React from 'react';

export interface PieChartProps {
    title: string;
    data: any;
    dataKey: string;
    dataName: string;
    // xAxisKey: string;
    colors: string[];
}

export const MyPieChart: React.FC<PieChartProps> = observer((props: PieChartProps) => {
    const { title, data, dataKey, dataName, colors } = props;

    return (
        <Col style={{ width: '30%', height: '100%' }}>
            <Title level={2} style={{ textAlign: 'center' }}>
                {title}
            </Title>
            <ResponsiveContainer width="100%" height="85%">
                <PieChart width={300} height={200} margin={{ top: 0, right: 0, bottom: 50, left: 0 }}>
                    <Pie
                        data={data}
                        cy={160}
                        innerRadius={70}
                        outerRadius={100}
                        fill="#8884d8"
                        paddingAngle={3}
                        dataKey={dataKey}
                        nameKey={dataName}
                    >
                        {data.map((entry, index) => (
                            <Cell key={`cell-${index}`} fill={colors[index % colors.length]} />
                        ))}
                    </Pie>
                    <Legend />
                    <Tooltip />
                </PieChart>
            </ResponsiveContainer>
        </Col>
    );
});
