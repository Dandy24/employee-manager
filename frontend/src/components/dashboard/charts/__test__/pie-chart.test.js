import { render } from '@testing-library/react';
import React from 'react';
import { MyLineChart } from '../line-chart';
import { MyPieChart } from '../pie-chart';

jest.mock('recharts', () => {
    const OriginalModule = jest.requireActual('recharts');

    return {
        ...OriginalModule,
        ResponsiveContainer: ({ height, children }) => (
            <OriginalModule.ResponsiveContainer width={800} height={500}>
                {children}
            </OriginalModule.ResponsiveContainer>
        ),
    };
});

const COLORS = ['rgb(136, 132, 216)', '#FFBB28', '#FF8042', 'rgb(130, 202, 157)'];

test('Pie chart matches snapshot', async () => {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(), // deprecated
            removeListener: jest.fn(), // deprecated
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });

    const testData = [
        { name: 'Hours worked', value: 1710 },
        { name: 'Hours vacation', value: 116 },
        { name: 'Hours sick', value: 48 },
        { name: 'Hours overtime', value: 52 },
    ];

    const view = render(
        <MyPieChart title="Rozložení hodin" data={testData} dataKey="value" dataName="name" colors={COLORS} />,
    );

    // eslint-disable-next-line testing-library/no-node-access,testing-library/no-container
    const pieChart = view.container.querySelector('#pie-chart');

    expect(pieChart).toMatchSnapshot();
});
