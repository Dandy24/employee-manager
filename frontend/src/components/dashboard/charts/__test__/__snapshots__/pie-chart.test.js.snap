// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pie chart matches snapshot 1`] = `
<svg
  class="recharts-surface"
  cx="50%"
  cy="50%"
  height="500"
  id="pie-chart"
  version="1.1"
  viewBox="0 0 800 500"
  width="800"
>
  <defs>
    <clippath
      id="pie-chart-clip"
    >
      <rect
        height="450"
        width="800"
        x="0"
        y="0"
      />
    </clippath>
  </defs>
  <g
    class="recharts-layer recharts-pie"
  >
    <g
      class="recharts-layer recharts-pie-sector"
    >
      <path
        class="recharts-sector"
        cx="400"
        cy="130"
        d="M 500,130
    A 100,100,0,
    1,0,
    462.8940023498477,207.74538229642548
  L 444.0258016448934,184.42176760749783
            A 70,70,0,
            1,1,
            470,130 Z"
        fill="rgb(136, 132, 216)"
        id="pie-chart-hours-worked-sector"
        name="Hours worked"
        stroke="#fff"
      />
    </g>
    <g
      class="recharts-layer recharts-pie-sector"
    >
      <path
        class="recharts-sector"
        cx="400"
        cy="130"
        d="M 466.8766872314382,204.34721719707062
    A 100,100,0,
    0,0,
    489.0462834421222,175.50559751442924
  L 462.33239840948556,161.85391826010047
            A 70,70,0,
            0,1,
            446.81368106200676,182.04305203794942 Z"
        fill="#FFBB28"
        id="pie-chart-hours-vacation-sector"
        name="Hours vacation"
        stroke="#fff"
      />
    </g>
    <g
      class="recharts-layer recharts-pie-sector"
    >
      <path
        class="recharts-sector"
        cx="400"
        cy="130"
        d="M 491.3058275657546,170.78291128073965
    A 100,100,0,
    0,0,
    496.41156156466906,156.5482729506465
  L 467.48809309526837,148.58379106545254
            A 70,70,0,
            0,1,
            463.91407929602826,158.54803789651777 Z"
        fill="#FF8042"
        id="pie-chart-hours-sick-sector"
        name="Hours sick"
        stroke="#fff"
      />
    </g>
    <g
      class="recharts-layer recharts-pie-sector"
    >
      <path
        class="recharts-sector"
        cx="400"
        cy="130"
        d="M 497.6688621217582,151.46609819787912
    A 100,100,0,
    0,0,
    499.8629534754574,135.23359562429442
  L 469.90406743282017,133.6635169370061
            A 70,70,0,
            0,1,
            468.3682034852307,145.02626873851537 Z"
        fill="rgb(130, 202, 157)"
        id="pie-chart-hours-overtime-sector"
        name="Hours overtime"
        stroke="#fff"
      />
    </g>
    <g
      class="recharts-layer recharts-pie-labels"
    >
      <g
        class="recharts-layer"
      >
        <path
          class="recharts-curve recharts-pie-label-line"
          cx="400"
          cy="130"
          d="M309.75200768165314,86.926807844001L291.7024092179838,78.31216941280121"
          fill="none"
          id="pie-chart-hours-worked-sector"
          name="Hours worked"
          points="[object Object],[object Object]"
          stroke="rgb(136, 132, 216)"
          type="linear"
        />
        <text
          alignment-baseline="middle"
          class="recharts-text recharts-pie-label-text"
          cx="400"
          cy="130"
          fill="rgb(136, 132, 216)"
          id="pie-chart-hours-worked-sector"
          name="Hours worked"
          stroke="none"
          text-anchor="end"
          x="291.7024092179838"
          y="78.31216941280121"
        >
          <tspan
            dy="0em"
            x="291.7024092179838"
          >
            88.79%
          </tspan>
        </text>
      </g>
      <g
        class="recharts-layer"
      >
        <path
          class="recharts-curve recharts-pie-label-line"
          cx="400"
          cy="130"
          d="M479.28400186920766,190.94298193888676L495.14080224304917,203.1315783266641"
          fill="none"
          id="pie-chart-hours-vacation-sector"
          name="Hours vacation"
          points="[object Object],[object Object]"
          stroke="#FFBB28"
          type="linear"
        />
        <text
          alignment-baseline="middle"
          class="recharts-text recharts-pie-label-text"
          cx="400"
          cy="130"
          fill="#FFBB28"
          id="pie-chart-hours-vacation-sector"
          name="Hours vacation"
          stroke="none"
          text-anchor="start"
          x="495.14080224304917"
          y="203.1315783266641"
        >
          <tspan
            dy="0em"
            x="495.14080224304917"
          >
            6.02%
          </tspan>
        </text>
      </g>
      <g
        class="recharts-layer"
      >
        <path
          class="recharts-curve recharts-pie-label-line"
          cx="400"
          cy="130"
          d="M494.1281614480112,163.76224551799152L512.9537937376134,170.51469462158983"
          fill="none"
          id="pie-chart-hours-sick-sector"
          name="Hours sick"
          points="[object Object],[object Object]"
          stroke="#FF8042"
          type="linear"
        />
        <text
          alignment-baseline="middle"
          class="recharts-text recharts-pie-label-text"
          cx="400"
          cy="130"
          fill="#FF8042"
          id="pie-chart-hours-sick-sector"
          name="Hours sick"
          stroke="none"
          text-anchor="start"
          x="512.9537937376134"
          y="170.51469462158983"
        >
          <tspan
            dy="0em"
            x="512.9537937376134"
          >
            2.49%
          </tspan>
        </text>
      </g>
      <g
        class="recharts-layer"
      >
        <path
          class="recharts-curve recharts-pie-label-line"
          cx="400"
          cy="130"
          d="M499.09882985299805,143.39484683624812L518.9185958235976,146.07381620349776"
          fill="none"
          id="pie-chart-hours-overtime-sector"
          name="Hours overtime"
          points="[object Object],[object Object]"
          stroke="rgb(130, 202, 157)"
          type="linear"
        />
        <text
          alignment-baseline="middle"
          class="recharts-text recharts-pie-label-text"
          cx="400"
          cy="130"
          fill="rgb(130, 202, 157)"
          id="pie-chart-hours-overtime-sector"
          name="Hours overtime"
          stroke="none"
          text-anchor="start"
          x="518.9185958235976"
          y="146.07381620349776"
        >
          <tspan
            dy="0em"
            x="518.9185958235976"
          >
            2.70%
          </tspan>
        </text>
      </g>
    </g>
  </g>
</svg>
`;
